### ISSUE TITLE
enhance link down handler to optimize cases where there is no alternative path

### ISSUE THREAD

**Issue by italovalcy** [kytos-ng/mef_eline#394]
Hi,

When a link down happens and affects an EVC that does not have alternative path, what happens is: kytos will remove the flows, try to find a new path, since there is no alternative path, pathfinder won't provide paths, and then the handle link down routine finishes with `[kytos.napps.kytos/mef_eline] [evc.py:740:deploy_to_path] (thread_pool_app_236) EVC(xxx, yyy) was not deployed. No available path was found. `

Later on, when the link comes UP, then a new Flow Removal will be triggered, a new path will be searched, and then finally the deploy will happens.

In those cases, one optimization that could be done is: if there is no path to handle the link down, we can leave the current_path as it is (ie., before trying to redeploy, we should check if redeployment is actually possible); when a link UP comes, let's check if the current_path could benefit from that link UP (i.e, no actions would be taken in the scenario above).

**Comment by viniarck** [comment]
@italovalcy, makes a lot of sense for dynamic EVCs, but in certain cases it would also imply that a worse path could end up being used for some time even though alternative best paths might become available again, so this is something to sort out if it's desirable or not, for instance:

- Suppose there's a dynamic EVC between SW1 and SW2 where three paths A, B, C can be found with total cost 10, 20, 30 respectively. Let's assume now, that initially it was using the best (lowest) path A, but then path A and B went down, and now there's C. So now, if a link goes down while using C and then we keep the `current_path` then if path A or B becomes available again it'll keep using C, which might be what you want or not.


```
+--------------+                         +---------------+
|              |          A (10)         |               |
|              |   -------------------   |               |
|     SW1      |          B (20)         |     SW2       |
|              |   --------------------  |               |
|              |          C (30)         |               |
|              |   --------------------  |               |
+--------------+                         +---------------+
```

Ultimately, this related side effect that I brought up is also related to static EVCs "switchover" from backup_path to primary_path once the primary_path is available again, which we also currently don't do. Anyway, depending on what you need, it might be worth thinking about a general path "optimizer" or "switchover" optimizer too, if it's ever desirable or not, maybe it's not desirable at all but bringing it up here to the discussion. Although overall high level optimizations might also be something for BAPM project too, but depending on roadmap simples ones still might make sense to implement on `mef_eline` directly too and enable them or not.

**Comment by italovalcy** [comment]
Thanks for the discussion, @viniarck, and for providing some examples and possible side effects! I was thinking more in the sense of EVCs that does not actually have alternative paths available. On the example you gave, since the EVC is now using C, if C goes down and A becomes available, we should switch back to A. If C goes down and A and B are also down, then we keep the state as if the EVC were using C. Then if C returns first, we just need to reactivate the status (no flows probably). ON the other hand, if A,B and C returns at the same time, then it will be subject to pathfinder handling multiple events simultaneously. (this is a challenge even the way it is now)

For EVCs with static primary and backup path, I believe mef eline already switch back to primary, no? https://github.com/kytos-ng/mef_eline/blob/master/models/evc.py#L1418 

We can first look into this optimization for EVCs that somehow need to request a path from pathfinder.

Please let me know if the comments above makes any sense.

**Comment by viniarck** [comment]
> On the example you gave, since the EVC is now using C, if C goes down and A becomes available, we should switch back to A. If C goes down and A and B are also down, then we keep the state as if the EVC were using C.

Right. The case I'm trying to point out though is, once it's on C and there's no alternative path, it won't move away from it unless there's a `link_down` event, so it might be using a non optimal path for a long time. 

> For EVCs with static primary and backup path, I believe mef eline already switch back to primary, no? https://github.com/kytos-ng/mef_eline/blob/master/models/evc.py#L1418

That's correct, my mistake, indeed switchover for static EVPLs is already working. The case that I described only apply to fully dynamic that won't try to deploy again after a [`link_up` if it already has a `current_path`](https://github.com/kytos-ng/mef_eline/blob/master/models/evc.py#L1425-L1426), which is also completely reasonable and save resources, but it can still keep using a non optimal path for a while.
